### PR TITLE
Test: Only sniff host metadata for node_selectors

### PR DESCRIPTION
**Pull request by nik9000**
Our rest testing framework has support for sniffing the host metadata on
startup and, before this change, it'd sniff that metadata before running
the first test. This prevents running these tests against
elasticsearch installations that won't support sniffing like Elastic
Cloud. This change allows tests to only sniff for metadata when they
encounter a test with a `node_selector`. These selectors are the things
that need the metadata anyway and they are super rare. Tests that use
these won't be able to run against installations that don't support
sniffing but we can just skip them. In the case of Elastic Cloud, these
tests were never going to work against Elastic Cloud anyway.

### CODE REVIEW

**Comment by nik9000** [comment]
I've intentionally opened this agianst 6.x because 6.x is where we actually use the node_selector. I'll forward port to master once reviewed and merged.

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java [hunk 1]
```diff
@@ -407,13 +395,16 @@ protected boolean randomizeContentType() {
     }
 
     /**
-     * Sniff the cluster for host metadata.
+     * Sniff the cluster for host metadata and return a
+     * {@link RestClientBuilder} for a client with that metadata.
      */
-    private List<Node> sniffHostMetadata() throws IOException {
+    protected final RestClientBuilder sniffHostMetadata() throws IOException {
```
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java]: I would rename this method to something like `getClientBuilderWithSniffedHosts`

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java [hunk 2]
```diff
@@ -65,19 +64,19 @@
...
-    private final CheckedConsumer<RestClientBuilder, IOException> clientBuilderConsumer;
+    private final CheckedSupplier<RestClientBuilder, IOException> clientBuilderBuilder;
 
     ClientYamlTestClient(
             final ClientYamlSuiteRestSpec restSpec,
             final RestClient restClient,
             final List<HttpHost> hosts,
             final Version esVersion,
-            final CheckedConsumer<RestClientBuilder, IOException> clientBuilderConsumer) {
+            final CheckedSupplier<RestClientBuilder, IOException> clientBuilderBuilder) {
```
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java]: can we rename the instance member and corresponding arguments to something like clientBuilderSupplier or clientBuilderWithSniffedNodes?

**Review by javanna — APPROVED** [review]
left a couple of nits on naming, LGTM otherwise. thanks @nik9000 !

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java (hunk 2 above)
Comment (nik9000) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ClientYamlTestClient.java]: Sounds good to me!

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java (hunk 1 above)
Comment (nik9000) [reply test/framework/src/main/java/org/elasticsearch/test/rest/yaml/ESClientYamlSuiteTestCase.java]: I'll do it!

**Pull request merged**